python/ray/data/_internal/logical/interfaces.py

```diff
@@ -0,0 +1,89 @@
+from typing import List
+
+
+class Operator:
+    """Abstract class for operators.
+
+    Operators are stateful and non-serializable; they live on the driver side of the
+    Dataset only.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["Operator"]):
+        self._name = name
+        self._input_dependencies = input_dependencies
+        for x in input_dependencies:
+            assert isinstance(x, Operator), x
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def input_dependencies(self) -> List["Operator"]:
+        """List of operators that provide inputs for this operator."""
+        assert hasattr(
+            self, "_input_dependencies"
+        ), "Operator.__init__() was not called."
+        return self._input_dependencies
+
+    def __repr__(self) -> str:
+        if self.input_dependencies:
+            out_str = ", ".join([str(x) for x in self.input_dependencies])
+            out_str += " -> "
+        else:
+            out_str = ""
+        out_str += f"{self.__class__.__name__}[{self._name}]"
+        return out_str
+
+    def __str__(self) -> str:
+        return repr(self)
+
+
+class LogicalOperator(Operator):
+    """Abstract class for logical operators.
+
+    A logical operator describes transformation, and later is converted into
+    physical operator.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["LogicalOperator"]):
+        super().__init__(name, input_dependencies)
+        for x in input_dependencies:
+            assert isinstance(x, LogicalOperator), x
+
+    def set_input_dependencies(self, input_dependencies: List["LogicalOperator"]):
+        self._input_dependencies = input_dependencies
+
+    def get_physical_dag(self):
```

Comment:
Shouldn't conversion of Logical->Physical plans be implemented by a separate planner class instead of an operator? Otherwise it kind of implies a 1:1 logical to physical mapping at least initially, which is a bit of an odd constraint.

Reply:
+1, in addition, this interface also feels like it's converting a single operator from logical to physical (rather than the entire DAG/QueryPlan).